### PR TITLE
Add all conversion codes

### PR DESCRIPTION
**Pull request by bendichter**

### CODE REVIEW

File: convert_towers_task/spikeglxdatainterface.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+"""Authors: Cody Baker and Ben Dichter."""
+from .baserecordingextractorinterface import BaseRecordingExtractorInterface
+from spikeextractors import SpikeGLXRecordingExtractor
+
+
+class SpikeGLXRecordingInterface(BaseRecordingExtractorInterface):
+    RX = SpikeGLXRecordingExtractor
```
Comment (bendichter) [inline convert_towers_task/spikeglxdatainterface.py]: This should be imported from NWB conversion tools

**Comment by CodyCBakerPhD** [comment]
This branch is meant simply as a primary location to add and debug all the separate modules; I was going to submit separate, far smaller PR's on a per module basis as they start getting filled in and debugged.

File: convert_towers_task/spikeglxdatainterface.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply convert_towers_task/spikeglxdatainterface.py]: Yeah, plan is to remove it here once that PR on nwb-conversion-tools is merged

**Comment by bendichter** [comment]
I think I merged that today

File: convert_towers_task/towersnwbconverter.py [hunk 2]
```diff
@@ -27,27 +29,34 @@ def get_metadata(self):
...
+                        datetime.strptime(date_text, "%Y/%m/%d") - timedelta(days=age)
+
         else:
+            subject_id = 'unknown'
             age = 'unknown'
             subject_data = {}
-            subject_data.update({'genotype': 'unknown'})
+            subject_data.update({'genotype': 'unknown',
+                                 'normWeight': 'unknown',
+                                 'date_of_birth': 'unknown'})
```
Comment (CodyCBakerPhD) [inline convert_towers_task/towersnwbconverter.py]: As a matter of style, dictionaries should always be specified as literals when possible.

File: convert_towers_task/towersnwbconverter.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                    subject_data['date_of_birth'] = \
+                        datetime.strptime(date_text, "%Y/%m/%d") - timedelta(days=age)
+
+        else:
+            subject_id = 'unknown'
+            age = 'unknown'
+            subject_data = {}
+            subject_data.update({'genotype': 'unknown',
+                                 'normWeight': 'unknown',
+                                 'date_of_birth': 'unknown'})
```
Comment (CodyCBakerPhD) [inline convert_towers_task/towersnwbconverter.py]:
```suggestion
            subject_data.update(
                genotype='unknown',
                normWeight='unknown',
                date_of_birth='unknown'
            )
```

File: convert_towers_task/utils.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            out = np.vstack((out, [tt[i + 1], tt[j]]))
+        out = np.vstack((out, [tt[before_jumps[-1] + 1], tt[-1]]))
+        return out
+    else:
+        return np.array([[tt[0], tt[-1]]])
+
+
+def mat_obj_to_dict(mat_struct):
+    """Recursive function to convert nested matlab struct objects to dictionaries."""
+    dict_from_struct = {}
```
Comment (CodyCBakerPhD) [inline convert_towers_task/utils.py]:
```suggestion
    dict_from_struct = dict()
```

File: convert_towers_task/towersnwbconverter.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                    subject_data['date_of_birth'] = \
+                        datetime.strptime(date_text, "%Y/%m/%d") - timedelta(days=age)
+
+        else:
+            subject_id = 'unknown'
+            age = 'unknown'
+            subject_data = {}
+            subject_data.update({'genotype': 'unknown',
+                                 'normWeight': 'unknown',
+                                 'date_of_birth': None})
```
Comment (bendichter) [inline convert_towers_task/towersnwbconverter.py]: age, genotype, weight, and data_of_birth are all optional fields in [pynwb.Subject](https://pynwb.readthedocs.io/en/stable/pynwb.file.html?highlight=Subject#pynwb.file.Subject), so there is no need to populate them with default arguments.

File: convert_towers_task/towerspositiondatainterface.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+from scipy.io import loadmat
+from datetime import timedelta
+from .utils import check_module, date_array_to_dt
+
+
+class TankPositionInterface(BaseDataInterface):
+    """Description here."""
+
+    @classmethod
+    def get_input_schema(cls):
```
Comment (bendichter) [inline convert_towers_task/towerspositiondatainterface.py]: why does this take an input arg?

File: convert_towers_task/towerspositiondatainterface.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            source_data=dict(
+                required=['folder_path'],
+                properties=dict(
+                    folder_path=dict(type='string')
+                )
+            )
+        )
+
+    def __init__(self, **input_args):
+        super().__init__(**input_args)
```
Comment (bendichter) [inline convert_towers_task/towerspositiondatainterface.py]: if all you are doing is calling super, this method doesn't need to be defined

File: convert_towers_task/towerspositiondatainterface.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        Parameters
+        ----------
+        nwbfile : NWBFile
+            DESCRIPTION.
+        metadata_dict : dict
+            DESCRIPTION.
+        stub_test : bool, optional
+            DESCRIPTION. The default is False.
+        include_spike_waveforms : bool, optional
+            DESCRIPTION. The default is False.
```
Comment (bendichter) [inline convert_towers_task/towerspositiondatainterface.py]: why is "include_spike_waveforms" in the position interface?

File: convert_towers_task/utils.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+from scipy.io import loadmat, matlab
+from dateutil.parser import parse as dateparse
+
+try:
+    from typing import ArrayLike
+except ImportError:
+    from numpy import ndarray
+    from typing import Union, Sequence
+    # adapted from numpy typing
+    ArrayLike = Union[bool, int, float, complex, list, ndarray, Sequence]
```
Comment (bendichter) [inline convert_towers_task/utils.py]: what's going on here? Is this for different versions of python?

File: convert_towers_task/utils.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def check_module(nwbfile, name, description=None):
```
Comment (bendichter) [inline convert_towers_task/utils.py]: maybe this could go in nwb-conversion-tools

File: convert_towers_task/utils.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    that are still matlab objects to dictionaries.
+    """
+    data = loadmat(mat_file_name, struct_as_record=False, squeeze_me=True)
+    for key in data:
+        if isinstance(data[key], matlab.mio5_params.mat_struct):
+            data[key] = mat_obj_to_dict(data[key])
+    return data
+
+
+def date_array_to_dt(array):
```
Comment (bendichter) [inline convert_towers_task/utils.py]: why not use `datetime.datetime`?

File: tank_lab_to_nwb/utils.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    Parameters
+    ----------
+    array : array of floats of the form [Y,M,D,H,M,S].
+
+    Returns
+    -------
+    datetime
+
+    """
+    temp = [str(round(x)) for x in array[0:-1]]
```
Comment (bendichter) [inline tank_lab_to_nwb/utils.py:91]:
```suggestion
    temp = [str(round(x)) for x in array[:-1]]
```

File: tank_lab_to_nwb/convert_towers_task/convert_towers.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                )
+            )
+
+            converter = TowersNWBConverter(**input_args)
+            metadata = converter.get_metadata()
+
+            # Session specific info
+            metadata['NWBFile'].update({'experimenter': experimenter})
+            metadata['NWBFile'].update({'session_description': paper_descr})
+            metadata['NWBFile'].update({'related_publications': paper_info})
```
Comment (bendichter) [inline tank_lab_to_nwb/convert_towers_task/convert_towers.py]:
```suggestion
            metadata['NWBFile'].update(
                experimenter=experimenter
                session_description=paper_descr,
                related_publications=paper_info
            )
```

File: tank_lab_to_nwb/convert_towers_task/convert_towers.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+            metadata['Subject'].update({'species': "Mus musculus"})
```
Comment (bendichter) [inline tank_lab_to_nwb/convert_towers_task/convert_towers.py]:
```suggestion
            metadata['Subject'].update(species="Mus musculus")
```

File: tank_lab_to_nwb/convert_towers_task/towersnwbconverter.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+            session_data = convert_mat_file_to_dict(mat_file_name=session_path)
+            subject_data = session_data['log']['animal']
+
+            for key in ['name', 'importAge', 'normWeight', 'genotype']:
+                if key not in subject_data.keys():
+                    subject_data[key] = 'unknown'
+
+            metadata['Subject'].update(subject_id=subject_data['name'])
+            metadata['Subject'].update(age=str(subject_data['importAge']))
+            metadata['Subject'].update(weight=str(subject_data['normWeight']))
```
Comment (bendichter) [inline tank_lab_to_nwb/convert_towers_task/towersnwbconverter.py]:
```suggestion
            metadata['Subject'].update(
                subject_id=subject_data['name'],
                age=str(subject_data['importAge']),
                weight=str(subject_data['normWeight'])
            )
```

File: tank_lab_to_nwb/convert_towers_task/towersnwbconverter.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+            # },
+            TowersPosition=dict()
+        )
+
+        if os.path.isfile(session_path+".mat"):
+            session_data = convert_mat_file_to_dict(mat_file_name=session_path)
+            subject_data = session_data['log']['animal']
+
+            for key in ['name', 'importAge', 'normWeight', 'genotype']:
+                if key not in subject_data.keys():
```
Comment (bendichter) [inline tank_lab_to_nwb/convert_towers_task/towersnwbconverter.py]:
```suggestion
                if key not in subject_data:
```

File: tank_lab_to_nwb/convert_towers_task/convert_towers.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+base_path = "D:/Neuropixels/TowersTask"
+# base_path = "/mnt/scrap/cbaker239/Neuropixels/TowersTask"
+
+# Manual list of selected sessions that cause problems with the general functionality
+exlude_sessions = []
+
+experimenter = ""
+paper_descr = ""
+paper_info = ""
```
Comment (bendichter) [inline tank_lab_to_nwb/convert_towers_task/convert_towers.py]: I'd rather not use empty strings. None of these are required fields, so if they are not present lets just leave them out

File: tank_lab_to_nwb/convert_towers_task/towersnwbconverter.py [hunk 18]
```diff
@@ -0,0 +1,92 @@
+"""Authors: Cody Baker and Ben Dichter."""
+from nwb_conversion_tools import NWBConverter
+from nwb_conversion_tools import SpikeGLXRecordingInterface
+from tank_lab_to_nwb.convert_towers_task.towerspositiondatainterface import TowersPositionInterface
```
Comment (bendichter) [inline tank_lab_to_nwb/convert_towers_task/towersnwbconverter.py]:
```suggestion
from .towerspositiondatainterface import TowersPositionInterface
```

File: tank_lab_to_nwb/convert_towers_task/towersnwbconverter.py (hunk 15 above)
Comment (bendichter) [reply tank_lab_to_nwb/convert_towers_task/towersnwbconverter.py]: age should be in ISO 8601 Duration format, e.g. 100 days only would be "P100D"

**Pull request merged**